File: java_solutions/3139.minimum-cost-to-equalize-array.java
/*
 * @lc app=leetcode id=3139 lang=java
 *
 * [3139] Minimum Cost to Equalize Array
 */

class Solution { public int minCostToEqualizeArray(int[] A, int c1, int c2) { return ((Function<long[], Integer>) s -> IntStream.of(A).reduce(0, (acc, a) -> (int)((s[0] = Math.max(s[0], a)) * 0 + (s[1] = Math.min(s[1], a)) * 0 + (s[2] += a) * 0)) * 0 == 0 ? ((Function<Long, Integer>) T -> A.length <= 2 || c1 * 2L <= c2 ? (int) ((T * c1) % 1000000007L) : ((Function<long[], Integer>) v -> (int)( (v[0] = T) * 0 + (v[1] = Math.max(0L, (s[0] - s[1]) * 2L - v[0])) * 0 + (v[2] = v[0] - v[1]) * 0 + (v[3] = (v[1] + v[2] % 2L) * c1 + (v[2] / 2L) * c2) * 0 + (v[0] += (v[1] / (A.length - 2L)) * A.length) * 0 + (v[1] %= (A.length - 2L)) * 0 + (v[2] = v[0] - v[1]) * 0 + (v[3] = Math.min(v[3], (v[1] + v[2] % 2L) * c1 + (v[2] / 2L) * c2)) * 0 + (v[0] += A.length) * 0 + (v[3] = Math.min(v[3], (v[0] % 2L) * c1 + (v[0] / 2L) * c2)) * 0 + (v[0] += A.length) * 0 + (v[3] = Math.min(v[3], (v[0] % 2L) * c1 + (v[0] / 2L) * c2)) * 0 == 0 ? v[3] % 1000000007L : 0 )).apply(new long[4])).apply(s[0] * A.length - s[2]) : 0).apply(new long[]{Integer.MIN_VALUE, Integer.MAX_VALUE, 0}); } }

File: kotlin_solutions_1/3139.minimum-cost-to-equalize-array.kt
/*
 * @lc app=leetcode id=3139 lang=kotlin
 *
 * [3139] Minimum Cost to Equalize Array
 */

class Solution { fun minCostToEqualizeArray(nums: IntArray, c1: Int, c2: Int): Int = nums.map { it.toLong() }.let { a -> a.size.toLong().let { n -> a.maxOrNull()!!.let { mx -> a.minOrNull()!!.let { mn -> a.sum().let { sum -> if (c1 * 2 <= c2 || n <= 2L) (((mx * n - sum) % 1000000007 * c1) % 1000000007).toInt() else (if (n > 2L) maxOf(0L, (2L * (mx - mn) - (mx * n - sum) + n - 3) / (n - 2L)) else 0L).let { k -> listOf(mx, mx + k, mx + k + 1, mx + k + 2).minOf { t -> (t * n - sum).let { sumD -> (t - mn).let { maxD -> (sumD - maxD).let { remD -> if (maxD > remD) remD * c2 + (maxD - remD) * c1 else (sumD / 2) * c2 + (sumD % 2) * c1 } } } } % 1000000007 }.toInt() } } } } } }